src/core/config/npo/n_plus_one.rs

```diff
@@ -1,63 +1,145 @@
 struct FindFanOutContext<'a> {
     config: &'a Config,
-    type_name: &'a String,
-    path: Vec<(String, String)>,
+    type_name: &'a str,
     is_list: bool,
 }
 
-fn find_fan_out(context: FindFanOutContext) -> Vec<Vec<(String, String)>> {
-    let config = context.config;
-    let type_name = context.type_name;
-    let path = context.path;
-    let is_list = context.is_list;
-    match config.find_type(type_name) {
-        Some(type_) => type_
-            .fields
-            .iter()
-            .flat_map(|(field_name, field)| {
-                let mut new_path = path.clone();
-                new_path.push((type_name.clone(), field_name.clone()));
-                if path
-                    .iter()
-                    .any(|item| &item.0 == type_name && &item.1 == field_name)
-                {
-                    Vec::new()
-                } else if field.has_resolver() && !field.has_batched_resolver() && is_list {
-                    vec![new_path]
-                } else {
-                    find_fan_out(FindFanOutContext {
+#[inline(always)]
+fn find_fan_out<'a>(
+    ctx: FindFanOutContext<'a>,
+    visited: &mut HashMap<TypeName<'a>, HashSet<FieldName<'a>>>,
+) -> YieldInner<'a> {
+    let config = ctx.config;
+    let type_name: TypeName = TypeName(ctx.type_name);
+    let is_list = ctx.is_list;
+    let mut ans = HashMap::new();
+
+    if let Some(type_) = config.find_type(type_name.0) {
+        for (field_name, field) in type_.fields.iter() {
+            let cur: FieldName = FieldName(field_name.as_str());
+            let tuple: (FieldName, TypeName) = (cur, TypeName(field.type_of.as_str()));
+
+            let condition = visited
+                .get(&type_name)
+                .map(|v: &HashSet<FieldName>| v.contains(&cur))
+                .unwrap_or_default();
+
+            if condition {
+                continue;
+            } else {
+                visited.entry(type_name).or_default().insert(cur);
+            }
+
+            if field.has_resolver() && !field.has_batched_resolver() && is_list {
+                ans.entry(type_name).or_insert(HashSet::new()).insert(tuple);
+            } else {
+                let next = find_fan_out(
+                    FindFanOutContext {
                         config,
                         type_name: &field.type_of,
-                        path: new_path,
                         is_list: field.list || is_list,
-                    })
+                    },
+                    visited,
+                );
+                for (k, v) in next.0 {
+                    ans.entry(k).or_insert(HashSet::new()).extend(v);
+                    ans.entry(type_name).or_insert(HashSet::new()).insert(tuple);
                 }
-            })
-            .collect(),
-        None => Vec::new(),
+            }
+        }
     }
+
+    YieldInner(ans)
 }
 
-pub fn n_plus_one(config: &Config) -> Vec<Vec<(String, String)>> {
+pub fn n_plus_one(config: &Config) -> Yield {
```

Comment:
Instead of directly exposing this function. Let's create a new struct called `NPOIdentifier` with a method called `identify` that returns the `Yield`